flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/allocator/SlotSharingSlotAllocator.java

```diff
@@ -133,7 +138,56 @@ public Optional<VertexParallelismWithSlotSharing> determineParallelism(
         return Optional.of(new VertexParallelismWithSlotSharing(allVertexParallelism, assignments));
     }
 
-    private static Map<JobVertexID, Integer> determineParallelism(
+    /**
+     * Distributes free slots across the slot-sharing groups of the job. Slots are distributed as
+     * evenly as possible while taking the minimum parallelism of contained vertices into account.
+     */
+    private static Map<SlotSharingGroupId, Integer> determineSlotsPerSharingGroup(
+            JobInformation jobInformation, int freeSlots) {
+        int numUnassignedSlots = freeSlots;
+        int numUnassignedSlotSharingGroups = jobInformation.getSlotSharingGroups().size();
+
+        final Map<SlotSharingGroupId, Integer> slotSharingGroupParallelism = new HashMap<>();
+
+        for (Tuple2<SlotSharingGroup, Integer> slotSharingGroup :
+                sortSlotSharingGroupsByUpperParallelism(jobInformation)) {
+            final int groupParallelism =
+                    Math.min(
+                            slotSharingGroup.f1,
+                            numUnassignedSlots / numUnassignedSlotSharingGroups);
```

Comment:
We are taking the min of the maximum parallelism of the SlotSharingGroup and the fair share of slots for each slot sharing group. I'm assuming if the adaptive scheduler does not run in reactive mode, each SlotSharingGroup will receive the desired number of slots. Only in reactive mode, this will lead to downscaling.

Reply:
> I'm assuming if the adaptive scheduler does not run in reactive mode, each SlotSharingGroup will receive the desired number of slots.

hmm that doesn't sound right. It _always_ depends on the number of free slots the scheduler has received.

If you have 2 vertices in separate SSGs with say p=4 and p=2 respectively, and get 3 slots from the RM, then one SSG/vertex gets 2, slots the other 1.

If you then
* lose a slot you downscale to 1 each.
* gain a slot you scale up to 2 each.
* gain 3 slots you scale up to 4/2 respectively.

Reply:
That is terrible from an autoscaler perspective, as we want full control over the scaling. We never want the adaptive scheduler to reduce the parallelism in any way. Instead, we provide the spec and the scheduler has to rescale the job safely. It either fulfils the rescaling request, or it does nothing. We do not have unlimited retries for rescaling. Rescaling is costly especially for stateful workloads.

Reply:
Sorry for the meta comment but I'm looking at this from the perspective of a future Rescale API.

Reply:
That said, all of this is not set in stone. We can probably add a mode to the adaptive scheduler which disables downscaling in combination with pre-requesting the right amount of resources.

Reply:
So you prefer the job not running at all then? Because that's the alternative.

Reply:
From the perspective of a Rescale API, I prefer the job to continue as-is, effectively ignoring the rescale call if the available resources are not available. The adaptive scheduler would need two modes: (a) reactive (b) declarative. We would use (a) for the reactive mode, (b) for the Rescale API.

Reply:
One assumption that we're making with the proposed rescale api is that when you use the API with the intent of scaling up to _exactly_ X then you must already _know_ that these resources can be made available to the job, be it either by spawning new TMs or down-scaling other jobs within a certain amount of time.

This would only not be the case for temporary issues (like say, a TM crash at just the right/wrong time) that would be remedied eventually, and thus could be covered by the stabilization timeout. (it also seems somewhat odd to handle this case so differently to when an in-use TM crashes and forces a down-scaling).

Anyhow, this is largely irrelevant to the PR, and I'd suggest to raise these points in the FLIP discussion. IIRC this also came up in our internal discussions and we concluded that this _may_ be achievable with a `lowerBound` parallelism, IFF the current job running below the minimum wouldn't be interrupted. Basically we consider this a potential follow-up.
"When" to _apply_ a rescaling (in terms of time, like shortly after a checkpoint, or in terms of value, aka cost of rescaling vs benefit) is a topic that we're quite interested in.

Reply:
Thanks for elaborating! I just saw [the proposal](https://cwiki.apache.org/confluence/display/FLINK/FLIP-291%3A+Externalized+Declarative+Resource+Management), I had somehow missed it before.

>Then you must already know that these resources can be made available to the job, be it either by spawning new TMs or down-scaling other jobs within a certain amount of time.

We are assuming a Kubernetes environment. It is possible that the k8s cluster is at max capacity; we will still need manual intervention in this case. However, we absolutely want to prevent jobs not recovering from a scaling operation due to lack of resources. If we could set resource requirements with min/max boundaries and those are guaranteed before restarting the job, then there won't be an issue in case of a resource-constrained environment because existing jobs continue to run. 

Considering the "when" and "how", Gyula and I have thought quite a bit about effective scaling. The initial tests we have run with customers have yielded promising results. We haven't optimized the checkpoint timing yet which is something we want to look into as well.

Reply:
Agreed, the declarative API should be powerful enough to allow users to express the trade-off they're willing to make.

As suggested, setting the "lower bound" = "upper bound" for parallelism will probably do the trick. Let's discuss is on the FLIP.